samples/hello-cube.html

```diff
@@ -8,35 +8,39 @@
 <body>
 <canvas height=1000 width=1000></canvas>
 <script>
-/*******
-* This preview of Web GPU uses WHLSL as its shading language.
-* The choice of shading language(s) that will be ingested by the API is still under deliberation.
-*******/
 const positionAttributeNum  = 0;
 const colorAttributeNum     = 1;
 const transformBindingNum   = 0;
 const bindGroupIndex        = 0;
+const colorLocation         = 0;
 
 const shader = `
 struct FragmentData {
-    float4 position : SV_Position;
-    float4 color : attribute(${colorAttributeNum});
-}
+    [[builtin(position)]] position: vec4<f32>;
+    [[location(${colorLocation})]] color: vec4<f32>;
+};
+
+[[block]]
+struct Uniforms {
+    modelViewProjectionMatrix: mat4<f32>;
+};
+
+[[group(${bindGroupIndex}), binding(${transformBindingNum})]] var<uniform> uniforms;
 
-vertex FragmentData vertex_main(
-    float4 position : attribute(${positionAttributeNum}),
-    float4 color : attribute(${colorAttributeNum}),
-    constant float4x4[] modelViewProjectionMatrix : register(b${transformBindingNum}))
-{
+[[stage(vertex)]]
+fn main(
+    [[location(${positionAttributeNum})]] position: vec4<f32>
+    [[location(${colorAttributeNum})]] color: vec4<f32>,
+) -> FragmentData {
     FragmentData out;
-    out.position = mul(modelViewProjectionMatrix[0], position);
+    out.position = mul(uniforms.modelViewProjectionMatrix[0], position);
     out.color = color;
 
     return out;
 }
 
-fragment float4 fragment_main(float4 color : attribute(${colorAttributeNum})) : SV_Target 0
-{
+[[stage(fragment)]]
+fn main([[location(${colorLocation})]] color: vec4<f32>) -> [[location(0)]] vec4<f32> {
```

Comment:
Doesn't this have to accept `FragmentData` as the input as the output from the vertex shader was a `FragmentData` struct (i.e. don't they have to match)? What happened to `position`?

Reply:
They don't have to match exactly. The matching rules are the same as before: the inputs for the next stage need to be a subset of the outputs of the previous stage, identified by locations.

Reply:
K, makes the matching more annoying as we'll have to dig through the struct, but should be fine.